NAS-103478 / 11.3 / Properly document Job and Event subscription methods for api v2 (#3675)
freenas/freenas: src/middlewared/middlewared/apidocs/templates/websocket/events.md

# Websocket Events

> Events are triggers which are generated under certain scenarios or at a certain period of time.

Reviewer: ```suggestion
Events are triggers that are generated under certain scenarios or at a certain period of time.
```

> Some events can accept arguments and then return results influenced by the passed arguments.

Reviewer: ```suggestion
Some events can accept arguments and return results that are influenced by those arguments.
```

> For passing arguments to the events, the following format is followed:

Reviewer: ```suggestion
Follow this format to pass arguments to events:
```

`event_name:arg`

> `arg` is parsed automatically by the event if it accepts it. For events which don't accept

Reviewer: ```suggestion
If  `arg` is accepted by the event, it is parsed automatically. Events that do not accept
```

> arguments, just the event name is used when subscribing to the event.

Reviewer: ```suggestion
arguments only use the event name when subscribing to the event.
```

{% for name, attrs in events.items() %}
### {{name}}
{% if attrs.description %}
#####  {{ attrs.description|markdown|safe }}
{% endif %}

{% if attrs.wildcard_subscription %}

> This event can be subscribed with wildcard `*` as event name.

Reviewer: ```suggestion
This event can be subscribed to with the wildcard `*` as the event name.
```

{% else %}

> This event cannot be subscribed with wildcard `*` as event name.

Reviewer: ```suggestion
This event cannot be subscribed to with the wildcard `*` as the event name.
```

{% endif %}

{% endfor %}

## Subscribing to Events

> Events are generated by the system based on certain conditions when they are met. It is not useful

Reviewer: ```suggestion
Events are generated by the system based on when certain conditions are met. It is not useful
```

> if the system is generating an event but there is no event listener to listen to the event. The

Reviewer: ```suggestion
if the system is generating an event and there is no event listener. Listening to
```

> mechanism of listening to events is termed as `subscribing` to an event.

Reviewer: ```suggestion
events is called `subscribing`.
```

> If a client wants to subscribe to all system events, they can specify `*` instead of the event name

Reviewer: ```suggestion
A client can subscribe to all system events by specifying `*`. This only applies to events that accept `*`
```

> and they would receive system events which accept `*` as a wildcard (please refer to list above

Reviewer: ```suggestion
as a wildcard (refer to the list above for events that accept `*`).
```

> for events which accept this and which don't) .

Reviewer: ```suggestion

```

### Websocket Client Subscription

Request:

    :::javascript
    {
        "id": "ad4dea8f-53a8-9a5c-1825-523e218c13ca",
        "name": "*",
        "msg": "sub"
    }

Response:

    :::javascript
    {
        "msg": "ready",
        "subs": ["ad4dea8f-53a8-9a5c-1825-523e218c13ca"]
    }
 

> Above example subscribes the websocket client to system events which accept a wildcard. 

Reviewer: ```suggestion
The example above subscribes the websocket client to system events that accept `*` as a wildcard. 
```

> Now each time an event is generated by the system, websocket client would get the event.

Reviewer: ```suggestion
Each time an event is generated by the system the websocket client would get the event.
```

Event Response Example:

    :::javascript
    {
        "msg": "changed",
        "collection": "core.get_jobs",
        "id": 79,
        "fields": {
            "id": 79, "method": "zfs.pool.scrub",
            "arguments": ["vol1", "START"], "logs_path": null,
            "logs_excerpt": null,
            "progress": {"percent": 0.001258680822502356, "description": "Scrubbing", "extra": null},
            "result": null, "error": null, "exception": null, "exc_info": null,
            "state": "RUNNING", "time_started": {"$date": 1571297741181},
            "time_finished": null
        }
    }

> The above event was generated by the system on a pool scrub.

Reviewer: ```suggestion
The event above was generated by the system when a pool is scrubbed.
```

> Following is an example for subscribing to `reporting.realtime` event:

Reviewer: ```suggestion
The example below is how to subscribe to the `reporting.realtime` event.
```

        "id": "8592f7c2-ce2b-4466-443a-80bbae5937d9",
        "name": "reporting.realtime",
        "msg": "sub"
    }

Response:

    :::javascript
    {
        "msg": "ready",
        "subs": ["8592f7c2-ce2b-4466-443a-80bbae5937d9"]
    }

Event Response Example:

    :::javascript
    {
        "msg": "added", "collection": "reporting.realtime",
        "fields": {
            "virtual_memory": {
                "total": 4784615424, "available": 854155264, "percent": 82.1,
                "used": 3779424256, "free": 136634368, "active": 894599168,
                "inactive": 717520896, "buffers": 0, "cached": 0,
                "shared": 188002304, "wired": 2884825088
            },
            "cpu": {"temperature": {}},
            "interfaces": {
                "em0": {
                    "received_bytes": 1068597254, "received_bytes_last": 1068597254,
                    "sent_bytes": 78087857, "sent_bytes_last": 78087857
                },
                "lo0": {
                    "received_bytes": 358364554, "received_bytes_last": 358364554,
                    "sent_bytes": 358360787, "sent_bytes_last": 358360787
                }
            }
        }
    }

> Following is an example on how we can subscribe to jobs:

Reviewer: ```suggestion
The example below is how to subscribe to jobs.
```

Request:

    :::javascript
    {
        "id": "19922f7c2-ce2b-4455-443a-80bbae5937a2",
        "name": "core.get_jobs",
        "msg": "sub"
    }

Response:

    :::javascript
    {
        "msg": "ready",
        "subs": ["19922f7c2-ce2b-4455-443a-80bbae5937a2"]
    }

Event Response Example:

    :::javascript
    {
        "msg": "added", "collection": "core.get_jobs", "id": 26,
        "fields": {
            "id": 26, "method": "jail.stop", "arguments": ["abc"],
            "logs_path": null, "logs_excerpt": null,
            "progress": {"percent": null, "description": null, "extra": null},
            "result": null, "error": null, "exception": null, "exc_info": null,
            "state": "WAITING", "time_started": {"$date": 1571305262662},
            "time_finished": null
        }
    }

> The above event was generated when a jail was stopped and a job for stopping the jail started.

Reviewer: ```suggestion
The event above was generated when a jail was stopped and a job for stopping the jail started.
```

> Event response shows that system has registered the job and it's waiting to be executed.

Reviewer: ```suggestion
The event response shows that system has registered the job and the job is waiting to be executed.
```
